File: src/main/java/org/lanternpowered/server/attribute/LanternAttributeModifierBuilder.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.attribute;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

public final class LanternAttributeModifierBuilder {

    private Double value;
    private LanternOperation operation;

    /**
     * How can you check whether the value scales between the min and max value
     * if there is no way to access the attribute?
     */
    public LanternAttributeModifierBuilder value(double value) throws IllegalArgumentException {
        this.value = value;
        return this;
    }

    public LanternAttributeModifierBuilder operation(LanternOperation operation) {
        this.operation = checkNotNull(operation, "operation");
        return this;
    }

    public LanternAttributeModifier build() {
        checkState(this.value != null, "value is not set");
        checkState(this.operation != null, "operation is not set");
        return new LanternAttributeModifier(this.operation, this.value);
    }

    public LanternAttributeModifierBuilder reset() {
        this.value = null;
        this.operation = null;
        return this;
    }
}

File: src/main/kotlin/org/lanternpowered/server/attribute/LanternAttributeModifierBuilder.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.attribute

import org.lanternpowered.api.attribute.AttributeModifier
import org.lanternpowered.api.attribute.AttributeModifierBuilder
import org.lanternpowered.api.attribute.AttributeOperation
import java.util.UUID

class LanternAttributeModifierBuilder : AttributeModifierBuilder {

    private var value: Double? = null
    private var operation: AttributeOperation? = null
    private var uniqueId: UUID? = null

    override fun value(value: Double): AttributeModifierBuilder = apply { this.value = value }
    override fun uniqueId(uniqueId: UUID): AttributeModifierBuilder = apply { this.uniqueId = uniqueId }
    override fun operation(operation: AttributeOperation): AttributeModifierBuilder = apply { this.operation = operation }

    override fun reset(): LanternAttributeModifierBuilder = apply {
        this.value = null
        this.operation = null
        this.uniqueId = null
    }

    override fun build(): AttributeModifier {
        val operation = checkNotNull(this.operation) { "The operation is not set" }
        val value = checkNotNull(this.value) { "The value is not set" }
        val uniqueId = this.uniqueId ?: UUID.randomUUID()
        return LanternAttributeModifier(uniqueId, operation, value)
    }
}
